lean/bitvec.lean

```diff
@@ -0,0 +1,375 @@
+    ⟨0, nat.pos_pow_of_pos n dec_trivial⟩
+
+  -- bitvectors have a zero, at every length
+  instance {n:ℕ} : has_zero (bitvec n) := ⟨bitvec.zero n⟩
+
+  @[simp]
+  lemma bitvec_zero_zero (x : bitvec 0) : x.val = 0 :=
+    begin
+      cases x,
+      { simp [nat.pow_zero, lt_one_is_zero] at x_property,
+        simp, assumption
+      }
+    end
+
+end zero
+
+section one
+
+  -- Create a bitvector with the constant one, doing so requires
+  -- the bitvector to have positive length.
+  protected
+  def one_of_pos_len (n: ℕ) {H : n > 0} : bitvec n :=
+    ⟨1, calc
+          1   < 2^1 : by dec_trivial_tac
+          ... ≤ 2^n : by apply (pow_le_pow_of_le_right (zero_lt_succ _) (succ_le_of_lt H))⟩
+
+  -- In pratice, it's more useful to allow 0 length bitvectors to have 1
+  -- as well. This leads to a special case where the 0-length bitvector
+  -- 1 is really just 0. This turns out to simplify things.
+  protected
+  def one (n:ℕ) : bitvec n :=
+    begin
+      cases n,
+      case nat.zero
+      { apply bitvec.zero },
+      case nat.succ
+      { apply bitvec.one_of_pos_len, apply zero_lt_succ }
+    end
+
+  instance {n:ℕ} : has_one (bitvec n)  := ⟨bitvec.one n⟩
+
+end one
+
+protected def cong {a b : ℕ} (h : a = b) : bitvec a → bitvec b
+| ⟨x, p⟩ := ⟨x, h ▸ p⟩
+
+section bitwise
+
+  -- A fixed width version of nat.bitwise
+  -- This applies `f` to each bit in tthe vectors.
+  def fin_bitwise (f: bool → bool → bool) : ℕ → ℕ → ℕ → ℕ
+    | 0 _  _:= 0
+    | (nat.succ w) x y :=
+      let xr := x.bodd_div2 in
+      let yr := y.bodd_div2 in
+      nat.bit (f xr.fst yr.fst) (fin_bitwise w xr.snd yr.snd)
+
+  -- A theorem that nat.bit is less than a power of two, when the input
+  -- is.
+  --
+  -- The implicit parameters are chosen so that apply is useful here.
+  theorem bit_lt_pow2  {w:ℕ} {b : bool} {x : ℕ} (h : x < 2^w)
+  : nat.bit b x < 2^(w+1) :=
+  begin
+    -- Simplify 2^(w+1)
+    simp [nat.add_succ, nat.pow_succ, nat.mul_comm _ 2, eq.symm (nat.bit0_val _)],
+    -- Split on b and simplify bit
+    cases b; simp [bit],
+    case ff { apply nat.bit0_lt h, },
+    case tt { apply nat.bit1_lt_bit0 h, }
+  end
+
+  theorem bitwise_lt (f: bool → bool → bool) (w:ℕ)
+  : ∀(x y : ℕ),  fin_bitwise f w x y < 2^w :=
+  begin
+    induction w with w p,
+    { intros, dec_trivial_tac, },
+    intros,
+    unfold fin_bitwise,
+    apply bit_lt_pow2 (p _ _),
+  end
+
+  def bitwise {w:ℕ} (f: bool → bool → bool) (x y : bitvec w) : bitvec w :=
+    ⟨fin_bitwise f w x.val y.val, bitwise_lt _ _ _ _⟩
+
+  def not {w:ℕ} (x: bitvec w) : bitvec w := ⟨2^w - x.val - 1,
+    begin
+      have xval_pos : 0 < x.val + 1,
+      { apply (succ_pos x.val) },
+      apply (sub_lt _ xval_pos),
+      apply pos_pow_of_pos,
+      apply (succ_pos 1)
+    end⟩
+
+  def and {w:ℕ} : bitvec w → bitvec w → bitvec w := bitwise band
+  def or  {w:ℕ} : bitvec w → bitvec w → bitvec w := bitwise bor
+  def xor {w:ℕ} : bitvec w → bitvec w → bitvec w := bitwise bxor
```

Comment:
After some more looking, I saw there are already nat bitwise operations.  I'd recommend using those instead in case Lean has some behind the scenes bindings.  For left and right shifts there are `nat.shiftl` and `nat.shiftr`.  For and, or, xor, there are `nat.land`, `nat.lor`, and `nat.lxor`.